ease-lab/vhive
# Function image for gg port. Still need working ingress. Clean merge

To test `gg` and `vhive` on the same VM-
Install `gg` first to avoid protobuf compile error:
1. Clone `vhive_port` branch of gg repo: `git clone -b vhive_port https://github.com/faromero/gg.git`
3. Run the commands at https://github.com/faromero/gg/#build-directions, starting with the `sudo apt-get install` command followed by `./fetch-submodules.sh`, etc. (I have not been testing gg using the Docker setup)
4. Set the indicated environment variables at https://github.com/faromero/gg/#environment-variables

Installing `vhive`:
1. Clone `gg_port_new` branch of vhive repo: `git clone -b gg_port_new https://github.com/faromero/vhive.git`
3. Run `stock-only` scripts at https://github.com/faromero/vhive/blob/main/docs/developers_guide.md#testing-stock-knative-setup-or-images
4. Run `kn service apply gg-port-0 --filename configs/knative_workloads/gg_port.yaml`
5. This should output the url: http://gg-port-0.default.192.168.1.240.sslip.io
6. You may test the connection to this container by running `curl -v -X POST http://gg-port-0.default.192.168.1.240.sslip.io`, if a connection is correctly established, it should respond with a JSON error

To test:
1. cd into `gg/tests/hello_world` directory
2. Run `gg init`
3. Run `gg infer make`
4. This is where you should run `gg force --jobs 1 --engine vhive=<url-or-ip> hello`, but you can set the vhive parameter equal to anything because I currently have the outputted url from the `kn service apply` command hardcoded into the execution engine.

Note: This last command should send a request to the server function in the deployed container, that compiles the test hello world program but instead it loops and gives an error that it failed to fetch dependencies, as shown below:
![gg-force-error](https://user-images.githubusercontent.com/63618501/129704242-6c4a7ae8-913f-4384-980d-1e9abe26e5ea.JPG)
Currently, running `gg force` outputs the request it sends. I would suggest diverting output to /dev/null or a file (I usually run it with `> output.txt`) so that only the errors print to the terminal.
You should be able to observe similar error outputs when you run `kubectl logs` on the gg-port-0 deployment pod's `queue-proxy` as to the one in the image below.
![Capture](https://user-images.githubusercontent.com/63618501/129699221-08aa6e9c-3ec7-49b4-b95c-88b86c41b5f7.JPG)
Another interesting thing to note is that when you run `kubectl logs` on the same pod's `user-container`, you'll notice that all the responses have status 200, as shown below:
![user-container-capture](https://user-images.githubusercontent.com/63618501/129703849-256fbed0-5937-48b0-a676-b93c36ab5d23.JPG)
Last Note: This error does not occur when you deploy the gg port function using Docker, and I have gotten this test to run successfully when I deployed it that way, so I believe this is something specific to knative. Specifically, I believe that there may be unexpected errors when the deployed server function attempts to communicate with the remote storage to fetch the data it needs as part of its execution (the thunk dependencies).
